fe/fe-core/src/main/java/com/starrocks/connector/hive/HiveConnectorScanRangeSource.java

```diff
@@ -0,0 +1,409 @@
+            hdfsScanRange.setFile_format(fileFormat.toThrift());
+            if (fileFormat.isTextFormat()) {
+                hdfsScanRange.setText_file_desc(fileDesc.getTextFileFormatDesc().toThrift());
+            }
+
+            if (attachment.dataCacheOptions != null) {
+                TDataCacheOptions tDataCacheOptions = new TDataCacheOptions();
+                tDataCacheOptions.setPriority(attachment.dataCacheOptions.getPriority());
+                hdfsScanRange.setDatacache_options(tDataCacheOptions);
+            }
+
+            TScanRange scanRange = new TScanRange();
+            scanRange.setHdfs_scan_range(hdfsScanRange);
+            scanRangeLocations.setScan_range(scanRange);
+
+            if (blockDesc.isPresent()) {
+                if (blockDesc.get().getReplicaHostIds().length == 0) {
+                    String message = String.format("hdfs file block has no host. file = %s/%s",
+                            remoteFileInfo.getFullPath(), fileDesc.getFileName());
+                    throw new StarRocksPlannerException(message, ErrorType.INTERNAL_ERROR);
+                }
+
+                for (long hostId : blockDesc.get().getReplicaHostIds()) {
+                    String host = blockDesc.get().getDataNodeIp(hostId);
+                    TScanRangeLocation scanRangeLocation = new TScanRangeLocation(new TNetworkAddress(host, -1));
+                    scanRangeLocations.addToLocations(scanRangeLocation);
+                }
+            } else {
+                TScanRangeLocation scanRangeLocation = new TScanRangeLocation(new TNetworkAddress("-1", -1));
+                scanRangeLocations.addToLocations(scanRangeLocation);
+            }
+            buffer.add(scanRangeLocations);
+        }
+
+    }
+
+    public Iterator<TScanRangeLocations> createScanRangeIterator(RemoteFileInfo remoteFileInfo) {
+        return new ScanRangeIterator(remoteFileInfo);
+    }
+
+    private void updateBackendSplitFile(RemoteFileInfo partition) {
+        ConnectContext connectContext = ConnectContext.get();
+        if (connectContext == null) {
+            return;
+        }
+        long hugeFileSize = connectContext.getSessionVariable().getConnectorHugeFileSize();
+        for (RemoteFileDesc fileDesc : partition.getFiles()) {
+            if (fileDesc.isSplittable()) {
+                backendSplitCount += (fileDesc.getLength() + hugeFileSize - 1) / hugeFileSize;
+            } else {
+                backendSplitCount += 1;
+            }
+        }
+        // if splits is small comparing to nodes, then better not let backend do split.
+        int nodes = connectContext.getAliveComputeNumber() + connectContext.getAliveBackendNumber();
+        backendSplitFile = (backendSplitCount > 2 * nodes);
+    }
+
+    private void updateIterator() {
+        if (!hasMoreOutput) {
+            return;
+        }
+        if (remoteFileInfoSource == null) {
+            init();
```

Comment:
why not `init()` in the `setup()`. This method has too much logic, we may try to simplify it asap

Reply:
Two reasons:
1. I'd like this operation to be as lazy as possible.
2. it's compatible with our UT cases.